File: Data Structures & Algorithms/binary-tree-diameter/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
        int res = 0;
        int maxLen = func(root, res);
        return res;
    }
    int func(TreeNode* root, int &res){
        if(!root) return 0;
        int leftHeight = func(root->left, res);
        int rightHeight = func(root->right, res);
        res = max(res, leftHeight+rightHeight);
        return 1+max(leftHeight, rightHeight);
    }
};

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def dfs(self, root: TreeNode, max_so_far):
        if not root:
            return
        if root.val >= max_so_far:
            self.count+=1
        max_so_far = max(max_so_far, root.val)
        self.dfs(root.left, max_so_far)
        self.dfs(root.right, max_so_far)

    def goodNodes(self, root: TreeNode) -> int:
        self.count = 0
        self.dfs(root, root.val)
        return self.count